### ISSUE TITLE
Should ScopeLinks require a Variable Declaration

### ISSUE THREAD

**Issue by rTreutlein** [opencog/atomspace#2343]
While fixing opencog/ure#33 i ran into the issue that a lot of code seems to be assuming that ScopeLinks (or BindLinks in this specific case) have proper Variable Declarations but as they are optional this can cause issues.
For now I have fixed this by generating Variable Declarations using the Variables class in the URE Rule code. Which is a local workaround. A real solution would be to generate this Variable Declaration whenever a ScopeLink without one is added to an AtomSpace such that in the AtomSpace every ScopeLink has a Variable Declaration. Or alternatively by just disallowing ScopeLinks without a Variable Declaration.
I think this might make a bunch of code simpler because you no longer have to worry about ScopeLinks without a Variable Declaration.

**Comment by linas** [comment]
The optional variable declarations was made as a convenience, to allow smaller, simpler constructions to be made.  Scope links, like the other pattern links, have two steps: a compile step, and a run step.  The compile step is generates and fills out various ad-hoc structures that are needed to make the run step got as fast as possible. 

One of the things that scope link does for the compile step is to search for and process variables, stuff them into various c++std::maps to make them easy for the run step to access them. So, yes, under the covers, a "variable declaration" is always created, and is always kept.  It's just not converted back into Atoms or stored in the AtomSpace.

I doubt your proposal would simplify anything -- the compile step is still needed, the crawls for free variables are still needed, the quote-unquote stuff is needed, as are all the alpha/beta conversion steps.  Those are the things generating the high complexity in the compile step.   There's nothing there that can be removed just because there's always a variable declaration.

In general I dislike having to create/use/store more atoms than necessary, in part because atoms are big, slow, take up a lot of RAM.  Avoiding them results in smaller RAM footprints, faster execution.

I don't understand why you consider a "local solution" to be a work-around. To me, it makes sense to have the URE to either always use variable declarations, or never use them, and sure, mixing two different styles for this can cause issues.

**Comment by linas** [comment]
I dunno. Automatically creating and inserting a variable declaration, when one is missing, is an interesting idea. It wouldn't simplify anything -- it could not be done until after the compile step, when the code has finished analyzing the link to figure out what's in it.

Suppose we did automatically generate variable declarations and inserted them as atoms.  Who does this benefit? I guess it helps with debugging: when a human looks at it, they could say something like "oh interesting, the code thinks that X is a variable". The run-step does not need this, the run step works directly with the C++ structures and utilities, rather than with the atoms.

**Comment by rTreutlein** [comment]
Hm since you seem to be suggestion we only use get_variables() and never get_vardecl() would it make sense to remove the get_vardecl() funktion? So nobody uses that accidentally when they should have use get_variables(). That way we could avoid issues the like the one i had.

**Comment by ngeiswei** [comment]
Maybe `get_vardecl()` could auto-generate the variable declaration using `get_variables().get_vardecl()` when that one is missing. And eventually `get_variables()` could be ubiquitously used and `get_vardecl()` removed.

There's one case when a variables object doesn't make much sense though, when the variable declaration is not fully formed, such as wholly or partially quoted, etc.

**Comment by linas** [comment]
The rules for C++ users who are accessing internal structures inside of C++ classes are very different than the rules for scheme/python users. It is expected that the C++ users are sophisticated systems programmers who know what they're doing, and are focused on performance, or are trying to solve some particularly difficult problem that can't be solved with naive manipulation of Atoms. In this case, some certain amount of complexity and non-orthogonality of the code is tolerated.  That's the nature of systems internals.

It sounded like you were suggesting changes that would become visible to ordinary, naive, unsophisticated scheme/python API users.  For these people, I want a "principle of least surprise".  In this case, when they put an atom into the atomspace, they can expect to get the same one back out (and not a different one, which now has variable declarations in it). 

Messing with atomspace contents is sort-of dangerous:  If user A emails user B with the email `(ScopeLink (...body..))`  but locally gets `(ScopeLink (Variable foo) (... body))`  and user B locally works with  `(ScopeLink (Variable bar) (... body))` then there are potential traps if the two are not fully alpha-convertible.  In place of email, think of a distributed app using zeromq or something like that. 

So basically, for C++, do what you need to do. For published AtomSpace contents, be very careful.

**Comment by linas** [comment]
Closing. I think this was a discussion, and all issues were resolved. Please re-open if there's more.
